### ISSUE TITLE
Faster reductions on GPU?

### ISSUE THREAD

**Issue by mtomassoli** [Theano/Theano#5507]
Let's say we have a 10000x1000 matrix `X` and we want to execute `X.sum(axis=-1)`. This is at least 10 times slower than `X1 + X2` on my GTX 970.
I've never written code for the GPU, but I guess you're using some kind of O(logn) algorithm for reducing each row of the matrix.
Wouldn't it be faster to let each core reduce a single row of the matrix in this case using the obvious O(n) algorithm you would use on the CPU?
If you're already doing this then where is the 10x slowdown coming from?

I'm asking here because I can never get answers to these kinds of question.

**Comment by khaotik** [comment]
Reduce isn't so trivial to parallelize/optimize on GPU. Due to multiple "cores" in GPU sharing instruction and local memory, thing just gets more complicated. Check [this](http://developer.download.nvidia.com/compute/cuda/1.1-Beta/x86_website/projects/reduction/doc/reduction.pdf) for example.

**Comment by mtomassoli** [comment]
Yes, but in the case I'm depicting above it should be much easier because there would be no sharing and no synchronization needed. It'd be almost as easy as an elementwise addition and almost as fast.

Each row would be accessed by one and only one core. Each core would just sum up all the elements of the row assigned to it.

**Comment by lamblin** [comment]
Which GPU back-end are you using?
Did you find a time difference when comparing both?
Are you using theano profiling, with `profiling.ignore_first_call=True` to get these numbers?

In any case, we are thinking of refactoring the reduction operations at some point.
